pkg/controller/controller_instance.go

```diff
@@ -355,27 +354,27 @@ func (c *controller) reconcileServiceInstance(instance *v1alpha1.ServiceInstance
 		return c.pollServiceInstanceInternal(instance)
 	}
 
-	// If there's no async op in progress, determine whether the checksum
-	// has been invalidated by a change to the object. If the instance's
-	// checksum matches the calculated checksum, there is no work to do.
-	// If there's an async op in progress, we need to keep polling, hence
-	// do not bail if checksum hasn't changed.
+	// If there's no async op in progress, determine whether there is a new
+	// generation of the object. If the instance's generation does not match
+	// the reconciled generation, then there is a new generation, indicating
+	// that changes have been made to the instance's spec. If there is an
+	// async op in progress, we need to keep polling, hence do not bail if
+	// there is not a new generation.
 	//
 	// We only do this if the deletion timestamp is nil, because the deletion
 	// timestamp changes the object's state in a way that we must reconcile,
-	// but does not affect the checksum.
+	// but does not affect the generation.
 	//
 	// Note: currently the instance spec is immutable because we do not yet
 	// support plan or parameter updates.  This logic is currently meant only
 	// to facilitate re-trying provision requests where there was a problem
 	// communicating with the broker.  In the future the same logic will
 	// result in an instance that requires update being processed by the
 	// controller.
-	if instance.Status.Checksum != nil && instance.DeletionTimestamp == nil {
-		instanceChecksum := checksum.ServiceInstanceSpecChecksum(instance.Spec)
-		if instanceChecksum == *instance.Status.Checksum {
+	if instance.Status.ReconciledGeneration != 0 && instance.DeletionTimestamp == nil {
+		if instance.Status.ReconciledGeneration == instance.Generation {
 			glog.V(4).Infof(
-				"Not processing event for ServiceInstance %v/%v because checksum showed there is no work to do",
+				"Not processing event for ServiceInstance %v/%v because reconciled generation showed there is no work to do",
```

Comment:
on each of the cases below where we stop processing and `return err` do we need to set ReconciledGeneration to Generation because we're stopped processing?

Reply:
The Controller does not give up on processing the resource when we `return err`. The Controller will retry the reconciliation later. We only stop processing the resource when we `return nil`. There is no logic currently in the code to handle when we stop retrying. That will need to be done as part of the work for Limits of Work [1].

[1] https://github.com/kubernetes-incubator/service-catalog/issues/1157

Reply:
I'm ok with leaving it for another PR, but it seems like there are errors that should cause a retry and errors that will never work and we know will never work.  Not sure we should be trying in those latter cases.